lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,182 @@
+require 'nokogiri'
+require 'htmlentities'
+
+class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
+  IGNORABLE_TAGS = %w[html head body].freeze
+
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  # http://www.456bereastreet.com/archive/201005/void_empty_elements_and_self-closing_start_tags_in_html/
+  VOID_TAGS = %w[area base br col command hr img input keygen link meta param source wbr].freeze
+
+  attr_reader :truncated_string,
+    :max_length,
+    :tail,
+    :ignored_levels
+
+  def initialize(options)
+    @html_coder = HTMLEntities.new
+
+    @max_length = options[:max_length]
+    @tail = options[:tail]
+
+    @truncated_string = ""
+    @closing_tags = []
+    @estimated_length = 0
+    @ignored_levels = 0
+  end
+
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    return if max_length_reached || ignorable_tag?(name)
+
+    # If already in ignore mode, go in deeper
+    if ignore_mode?
+      enter_ignored_level unless single_tag_element?(name)
+      return
+    end
+
+    string_to_add = opening_tag(name, attributes)
+
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      enter_ignored_level unless single_tag_element?(name)
+      return
+    end
+
+    # Save the tag so we can push it on at the end
+    @closing_tags.push name unless single_tag_element?(name)
+
+    append_to_truncated_string(string_to_add, length_of_tags)
+  end
+
+  # This method is called when the parser encounters characters between tags
+  def characters(decoded_string)
+    return if max_length_reached || ignore_mode?
+
+    # Use encoded length, so &gt; counts as 4 bytes, not 1 (which is what '>' would give)
+    encoded_string = @html_coder.encode(decoded_string, :named)
+    string_to_append = if encoded_string.bytesize > remaining_length
+      # This is the line which prevents HTML entities getting truncated - treat them as a single char
+      str = @html_coder.encode(truncate_string(decoded_string), :named)
+      str << tail if remaining_length >= tail.bytesize
```

Comment:
Shouldn't this be checking `remaining_length - str.bytesize >= tail.bytesize` since `remaining_length` doesn't contain the new (truncated) string yet?